### PR TITLE
feat: ForkChoiceSubscriptions for BlockchainProvider && NoopProvidor

### PR DESCRIPTION
**Pull request by royvardhan**
Ref [9521](https://github.com/paradigmxyz/reth/issues/9521)

### CODE REVIEW

File: crates/storage/provider/src/providers/mod.rs [hunk 1]
```diff
@@ -667,7 +667,7 @@ where
 
         if let Some(block) = self.tree.pending_block_num_hash() {
             if let Ok(pending) = self.tree.pending_state_provider(block.hash) {
-                return self.pending_with_provider(pending)
+                return self.pending_with_provider(pending);
```
Comment (emhane) [inline crates/storage/provider/src/providers/mod.rs:670]: please revert these unrelated changes. use the nightly formatter for reth `cargo +nightly fmt`. you can also use shortcut `make lint`.

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/blockchain-tree/src/blockchain_tree.rs
Comment (mattsse) [inline crates/blockchain-tree/src/blockchain_tree.rs:1]: please undo the extra `;`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/blockchain-tree/src/blockchain_tree.rs
Comment (royvardhan) [reply crates/blockchain-tree/src/blockchain_tree.rs:1]: Done

**Review by mattsse — CHANGES_REQUESTED** [review]
hmm, after looking at this again, I'd like to take one step back and only implement this for `BlockChainProvider` and Noop.

blockchainprovider then should use these channels to obtain the listeners, 
https://github.com/paradigmxyz/reth/blob/e907f0eab0d90f286f2197ee8935f10817ff1edf/crates/chain-state/src/chain_info.rs#L139-L144

which means we also need to change this type slightly

https://github.com/paradigmxyz/reth/blob/e907f0eab0d90f286f2197ee8935f10817ff1edf/crates/chain-state/src/notifications.rs#L146-L146

**Pull request closed without merging**